File: Application/src/main/java/com/shareyourproxy/api/rx/command/GenerateShareLinkCommand.java
package com.shareyourproxy.api.rx.command;

import android.app.Service;
import android.os.Parcel;

import com.shareyourproxy.api.domain.model.GroupToggle;
import com.shareyourproxy.api.domain.model.User;
import com.shareyourproxy.api.rx.RxShareLink;
import com.shareyourproxy.api.rx.command.eventcallback.EventCallback;

import java.util.ArrayList;


/**
 * Generate public link urls.
 */
public class GenerateShareLinkCommand extends BaseCommand {
    public static final Creator<GenerateShareLinkCommand> CREATOR =
        new Creator<GenerateShareLinkCommand>() {
            @Override
            public GenerateShareLinkCommand createFromParcel(Parcel in) {
                return new GenerateShareLinkCommand(in);
            }

            @Override
            public GenerateShareLinkCommand[] newArray(int size) {
                return new GenerateShareLinkCommand[size];
            }
        };
    private final static java.lang.ClassLoader CL =
        GenerateShareLinkCommand.class.getClassLoader();

    public final ArrayList<GroupToggle> groups;
    public final User user;

    public GenerateShareLinkCommand(User loggedInUser, ArrayList<GroupToggle> groups) {
        this.groups = groups;
        this.user = loggedInUser;
    }

    private GenerateShareLinkCommand(Parcel in) {
        this((User) in.readValue(CL), (ArrayList<GroupToggle>) in.readValue(CL));
    }

    @Override
    public EventCallback execute(Service service) {
        return RxShareLink.INSTANCE.getShareLinkMessageObservable(service, user, groups);
    }

    @Override
    public int describeContents() {
        return 0;
    }

    @Override
    public void writeToParcel(Parcel dest, int flags) {
        dest.writeValue(user);
        dest.writeValue(groups);
    }


}

File: Application/src/main/kotlin/com/shareyourproxy/api/rx/command/GenerateShareLinkCommand.kt
package com.shareyourproxy.api.rx.command

import android.content.Context
import android.os.Parcel
import android.os.Parcelable
import com.shareyourproxy.api.domain.model.GroupToggle
import com.shareyourproxy.api.domain.model.User
import com.shareyourproxy.api.rx.RxShareLink.getShareLinkMessageObservable
import com.shareyourproxy.api.rx.command.eventcallback.EventCallback
import java.util.*


/**
 * Generate public link urls.
 */
internal final class GenerateShareLinkCommand(val user: User, val groups: ArrayList<GroupToggle>) : BaseCommand() {
    @Suppress("UNCHECKED_CAST")
    private constructor(parcel: Parcel) : this(parcel.readValue(CL) as User, parcel.readValue(CL) as ArrayList<GroupToggle>)

    override fun execute(context: Context): EventCallback {
        return getShareLinkMessageObservable(context, user, groups)
    }

    override fun writeToParcel(dest: Parcel, flags: Int) {
        dest.writeValue(user)
        dest.writeValue(groups)
    }

    companion object {
        private val CL = GenerateShareLinkCommand::class.java.classLoader
        val CREATOR: Parcelable.Creator<GenerateShareLinkCommand> = object : Parcelable.Creator<GenerateShareLinkCommand> {
            override fun createFromParcel(parcel: Parcel): GenerateShareLinkCommand {
                return GenerateShareLinkCommand(parcel)
            }

            override fun newArray(size: Int): Array<GenerateShareLinkCommand?> {
                return arrayOfNulls(size)
            }
        }
    }


}
